### PR TITLE
Geo-referenced / un-tiled raster sources

### PR DESCRIPTION
**Pull request by bcamper**
This PR adds support for un-tiled, geo-referenced raster image data sources, similar to un-tiled vector sources, but with the need to explicitly specify the geographic bounds of the image(s). This functionality is commonly used for overlays of historical images, local aerial imagery, etc.

**Basic Syntax**
The basic syntax is similar to `GeoJSON`/`TopoJSON` sources, where the *lack* of an `{x}`/`{y}`/`{z}` tile pattern in the URL indicates the image should be treated as a standalone/un-tiled source:

```
sources:
  chelan:
    type: Raster # can now indicate either tiled or un-tiled image source
    url: images/chelan.jpg # no tile XYZ pattern in URL indicates standalone image
    bounds: [-120.588, 47.467, -119.904, 48.033] # geo-extent of the image
```

This example places a historical map in the Chelan, WA area (with Walkabout base map):

screen shot 2018-12-30 at 1 55 59 pm

For this implementation, only simple rectangular images that are already oriented "North-up" and with a single rectangular bounding box are supported; future iterations could support arbitrary shapes or rotations with control points.

Because this branch also includes changes to build tooling and testing (see below) that creates a lot of noise in the diff, a more accurate diff of the raster functionality is here: https://github.com/tangrams/tangram/compare/5d0aaf...552b06

**Client-Side Tiling Implementation**
Un-tiled raster sources are implemented with a similar pattern as with un-tiled vector sources: the image is loaded and then tiled client-side into individual raster tiles (per zoom). Each raster tile is simply an appropriate sub-rectangle of the original image (scaled as necessary for the zoom). While I initially considered simply creating a single quad/texture for these images, tiling client-side has some advantages, especially in terms of integration with Tangram's existing processing pipeline:

- Using raster tiles internally means these sources work out-of-the-box with existing raster source functionality for sampling rasters inside shaders (`sampleRaster()` etc. functions), attaching them to other data sources (with `rasters: [...]`), and clipping against vector sources.
- Image quads would often extend past the range supported by our existing `position` attributes, which only span a few tiles (necessitating some support for additional VBO layouts and position vertex shader handling).
- Some overlay images can be quite large, and tiles allow for less GPU texture memory consumption, only loading what is needed based on the current viewport (especially as these images can be zoomed in or out quite far).

As with other data sources, the `bounds` parameter is used to determine if a given tile is needed for the source, so areas outside the raster image are skipped entirely.

**Alpha Test**
Alpha values, either in the raster image itself, or in the client-side-created tiles around the image's border, can be handled with regular blending in non-`opaque` blend modes without any additional logic. Since `opaque` mode does not handle alpha blending, an [additional alpha test](https://github.com/tangrams/tangram/blob/384056420ae3bbc72bda311b4f2b9339beb77ec9/src/styles/polygons/polygons_fragment.glsl#L56-L69) is used in the fragment shader to hide pixels with an alpha value less than 1 (I experimented with other thresholds such as 0.5, but requiring full alpha led to the best quality in avoiding antialiasing and other artifacts around the image border). 

**Additional Source Parameters**
For some applications, it is necessary to view many individual raster images at once. Keeping each image as a separate raster data source is not efficient for Tangram's pipeline, either in terms of internal implementation (e.g. a tile instance per data source) or scene syntax (e.g. styling multiple data sources is unwieldy as a layer can only draw from a single data source). To serve these cases, additional syntax for combining *multiple images into one raster data source* was added, with a `composite` parameter; this parameter takes an array of objects, each with its own `url` and `bounds` parameters. For example:

```
  images:
    type: Raster
    composite:
      - { url: image1.png, bounds: [...] }
      - { url: image2.png, bounds: [...] }
      - { url: image3.png, bounds: [...] }
      ...
```

Here's an example of a large collection of aerial drone images combined into a single raster source, where each field is a separate image:

screen shot 2018-12-31 at 4 52 47 pm

The `composite` parameter for multiple images is mutually exclusive with the `url` and `bounds` parameters for a single image, and will generate an error. I was unsure about adding a separate parameter for multiple images. An alternative would be to have the `url` and `bounds` parameters accept arrays, which would correspond to each other "in parallel". This seemed potentially less intuitive and more susceptible to error than an explicit parameter, but is also a reasonable option.

Finally, to allow for alpha values to be applied between these composited images, a data source-level `alpha` parameter is provided. This value can be set either for the entire data source, or per image within the `composite` array (with the latter taking precedence). The alpha is applied CPU-side, at the time of client-side tiling (this is necessary to allow for composited images to blend together). For example:

```
  image:
    type: Raster
    alpha: 0.7 # apply 70% alpha to images
    composite:
      - { url: ..., bounds: ... }
      - { url: ..., bounds: ... }
      - { url: ..., bounds: ..., alpha: 1 } # override to apply full alpha to this image
```

Multiple images with alpha applied:
screen shot 2018-12-31 at 5 03 09 pm

**Build & Test Tooling Changes**
As part of this development, support for `async/await` syntax -- which made handling of async tasks such as image loading and cross-thread calls to create tile textures more concise -- was added via the [`fast-async`](https://github.com/MatAtBread/fast-async) Babel plugin using [`nodent`](https://github.com/MatAtBread/nodent). Because the plugin can transpile these keywords with pure Promises, there's no overhead added to the ES5 build (the ES6 build already supports `async`/`await` natively without transpiling).

After adding `async/await` support I found out that `jshint` does not support it (or other newer JS syntax). This necessitated switching the linter to [`eslint`](https://eslint.org/), which we have already been planning to do (#678). A significant portion of the PR diff is from [this change](https://github.com/tangrams/tangram/commit/8cd70229e2e2dad5e5c34caffd544274e2abe167), which includes many automatic syntax changes, along with manual adjustments. For now, we are using eslint's default [recommended rules](https://github.com/tangrams/tangram/blob/384056420ae3bbc72bda311b4f2b9339beb77ec9/.eslintrc.js) with minor config for Tangram-specific style; in the future may adopt/extend an existing community standard eslint config.

Lastly, while testing this branch, the test suite was repeatedly failing when executed by Sauce Labs, inexplicably not loading Leaflet before executing the test suite. This behavior had intermittently been seen on other branches including `master`, but was persistent in this case, though there is no clear link to the changes in this branch. While debugging this, I upgraded all test packages (which did not solve the problem), and ultimately tried using vanilla Chrome on CircleCI, instead of using the tunnel to Sauce Labs with Firefox. This appears to fix the problem, and Chrome now appears to properly support WebGL tests in this context! This allowed us to drop the Sauce and custom Firefox launcher dependencies, simplifying the overall test pipeline (see https://github.com/tangrams/tangram/commit/384056420ae3bbc72bda311b4f2b9339beb77ec9).

### CODE REVIEW

**Comment by bcamper** [comment]
cc @meetar @matteblair

**Comment by bcamper** [comment]
@burritojustice @straup may also be relevant to your interests

File: src/camera.js [hunk 1]
```diff
@@ -30,7 +30,7 @@ export default class Camera {
     }
 
     // Called once per frame per program (e.g. for main render pass, then for each additional pass for feature selection, etc.)
-    setupProgram(program) {
+    setupProgram(/*program*/) {
```
[the remark this answers is not in the document]
Comment (bcamper) [reply src/camera.js]: Oh, this method is empty in the parent class, and needs to be defined by subclasses. The comment here was just to document the method arguments, without causing a lint error. There are other (maybe better) ways to dot hat - could use an eslint line exception comment instead.

**Comment by bcamper** [comment]
@meetar:
> So to confirm: if you combine multiple overlapping rasters in a composite source, you're able to apply alpha to the whole combined source without the rasters showing through each other, correct?

Maybe I wasn't clear (har har). When you have multiple overlapping rasters, the `alpha` is applied to each image, e.g. if `alpha: 0.5`, image A is drawn with 50% opacity, then image B is drawn over it with 50% opacity, etc., meaning you would see the interaction between the two (as in the screenshot example). If you set a *single* `alpha` value for the data source, that value applies to each image. Additionally, you can override the `alpha` for individual images within the `composite` list, which overrides the general alpha (if there is one); this was included in case you wanted to vary the alpha for different images. But in either case, *how* the alpha is applied is the same, regardless of *where* the alpha was set. Does that make sense?

If you *want* the individual images to occlude each other when drawing multiple composited rasters with alpha, but you *want* , you would set no alpha (or `alpha: 1`) for the raster source, and then set an alpha at the `draw` level, which would apply to the entire, composited raster layer. The differences between these two methods exposes some of the implementation internals of drawing at CPU vs. GPU level, but it served a necessary use case (being able to have rasters with interacting alpha).

**Comment by matteblair** [comment]
The syntax for this feature makes sense to me, I think. I'm a little unclear on the process of compositing multiple image sources though.

> The alpha is applied CPU-side, at the time of client-side tiling (this is necessary to allow for composited images to blend together).

@bcamper Can you elaborate on this? (regarding the source-level alpha)

**Comment by bcamper** [comment]
@matteblair @meetar ha OK clearly this one has caused confusion :)

I didn't initially implement the "source" alpha, but ended up doing so after realizing I couldn't serve all the necessary use cases without it. Here are 4 examples to show the different configurations.

The images are composited into a single raster tile layer, meaning no matter how many images you put in `composite`, there will only be one raster tile they are rendered into for any given point in space. This operation happens CPU-side, e.g. all the images that intersect the tile are drawn into it, and the final image is turned into a GL texture and uploaded to the GPU. In Tangram JS, this is done with the Canvas `drawImage()` API; in ES, I assume it could use a similar image library. 

If no alpha was applied at the time of compositing, then overlapping images with alpha will have an "x-ray" effect where they overlap, and the top-most image will "see through" any other images that were underneath it, through to the background map (see example 4). This is a similar effect to our `blend: translucent` for geometry. But probably the more common case is that you want to have the ability to blend alpha both *between* the images *and* with the background map (see example 2). Similarly, sometimes you may want to apply different alpha levels to images individually (see example 3). To accomplish this, we need to apply alpha blending to the images as they are composited, so that both their color and alpha values are blended in the texture data.

**Full Alpha**
Everything's opaque, the two raster images occlude each other and the background.

```
sources:
  images:
    type: Raster
    alpha: 1 # or omit this, full opacity is the default
    composite:
      - { url: ..., bounds: ... }
      - { url: ..., bounds: ... }
```
full-alpha

**Source Alpha: Same Value for All Images**
The source images are composited, with the same `alpha: 0.7` value applied to them. You can see here that this allows the images' alpha to interact both between them, and with the background, e.g. you can see the bottom image underneath the top one where they overlap (e.g. contour lines in the lake), and the background back behind both of them (e.g. terrain behind the larger image border). 

```
sources:
  images:
    type: Raster
    alpha: 0.7 # all images have 70% alpha applied
    composite:
      - { url: ..., bounds: ... }
      - { url: ..., bounds: ... }
```
source-alpha

**Source Alpha: Mixed Values**
The source images are composited, with a `alpha: 0.7` value for one and full `alpha: 1` opacity for the other. You can see that the top image is fully opaque, but the raster image underneath it still shows through to the map background.

```
sources:
  images:
    type: Raster
    alpha: 0.7 # default value for images
    composite:
      - { url: ..., bounds: ... }
      - { url: ..., bounds: ..., alpha: 1 } # override to set this image to full alpha
```
source-alpha-mixed

**Draw Alpha**
In this example, no alpha is applied during image compositing. The texture data will be fully opaque. But, we can still apply an alpha value at render time. This will apply alpha to the  entire raster layer, which is then draw semi-transparently against the map background. You can see here that the top image has an "x-ray" effect, seeing through to the map background, but without any of the bottom raster image visible (e.g. the contour lines in the lake are no longer visible, because they were coming from the bottom image, but the terrain shading coming from the background is still visible on the top image's border).

Note that this can only be done when using some form of alpha blending (`overlay`, `translucent`, etc.) for the draw style.

```
sources:
  images:
    type: Raster
    alpha: 1 # or omit this, full opacity is the default
    composite:
      - { url: ..., bounds: ... }
      - { url: ..., bounds: ... }
styles:
  raster-overlay: # draw rasters with overlay blending
    base: raster
    blend: overlay
layers:
  raster:
    data: { source: images }
    draw:
      raster-overlay:
        color: [1, 1, 1, 0.7] # apply alpha at draw time, to a fully opaque raster layer
```
draw-alpha

**Comment by matteblair** [comment]
Iiiinteresting. I'm glad for this lengthy explanation now because this is not what I expected! In short: Tangram ES should be able to match this behavior but would almost certainly use a very different implementation.

I don't think Tangram ES will composite source images on the CPU. I'm not aware of a suitable c/c++ library that provides this but isn't massive (e.g. OpenCV), and it takes a lot to justify the build time and binary size of a new dependency. But more importantly, it just seems like a lot of computation to do for each tile.

Instead we could render the images directly to a texture in GL - but that still results in a large number of textures to hold in memory. 

What seems like the most efficient approach is:
1. Generate geometry for the bounds of each image
2. Split the geometry into tiles
3. Assign UV coordinates to the geometry
4. For each tile, draw the geometries with the appropriate texture in their order in the composite list.

I feel like you must have considered this approach, so perhaps I'm overlooking something.

**Comment by meetar** [comment]
@matteblair With the approach you suggest, would you still be able to do the "Draw Alpha" example (aka precomposed alpha)? Or is that the same as the "cpu compositing" you were mentioning?

**Comment by bcamper** [comment]
@matteblair Honestly, it just seemed easier to implement it this way in JS, but I think your approach is a viable alternative that we could try in the future for comparison. Initially I considered just creating one big quad for each image, but this seemed to have a few issues, including attribute position precision, because some of the images can get quite large in space, and more importantly, didn't seem compatible with our terrain raster setup, where an image could span many underlying terrain tiles (more than the # of available GL samplers).

Your suggestion doesn't have those issues though and is like a middle-ground between one big quad, and the solution I used here, e.g. splitting the geometry into tiles, but not the raster itself. A few things I was unsure about with that route:

- Some of these textures (for historical maps for instance) can be quite large, and could exceed the maximum GL texture size (especially on lower end devices -- granted my iPhone 6 is old, but I have had problems with textures larger than 2048x2048).
- There can be a lot of textures composited at once (I have a use case with 280 drone images). I figured having a smaller, fixed / maximum number of GL textures (corresponding to tiles) could be more reliable. But maybe the difference between CPU and GPU memory is more blurred these days anyway? It also requires a lot of texture re-binding during drawing.
- Similar to the above, in zooming the map, you can end up seeing only a small section of the viewport (at high zoom), or the entire image scaled down to several times smaller than 100% (at low zooms). When the images are tiled, you know consistently how much memory you're consuming.

These may not actually be significant issues though. 

Regarding doing a C/C++ CPU composite, maybe I'm missing something, but it should just be a simple sub-rectangle copy, with some additional math to apply alpha / blending between sources? Once the image is decoded to raw pixel data, it's a fairly straightforward pixel operation per pixel, no? (I don't particularly mind if the implementations diverge, just curious.)

**Comment by matteblair** [comment]
@meetar The effect in the "draw alpha" example is the same as Tangram's "translucent" blend mode, so I expect we'd be able to implement it the same way while rendering separate textured quads.

@bcamper The point about images exceeding the hardware texture limit is a good one - I hadn't thought of that and it could be tricky to work around. Compositing images in software shouldn't be complicated in our case (with only one pixel format to consider) it just feels like it could take a noticeable amount of time without some kind of hardware acceleration - though maybe my estimation is wrong here. 

In any case, this looks like a solid feature! I think the set of options makes sense and Tangram ES should be able to support the same capabilities. No promises about a timeline for support though :)

**Comment by bcamper** [comment]
Thanks all! I will get this merged now so we can move towards a v0.17.0 release soon, but will consider trying the single-texture approach in the future (with caveats around max-texture-size... could perhaps still split those very large textures CPU-side into a few), especially if performance is an issue. FWIW, the CPU compositing has not seemed to take long (but is undoubtedly slower).

**Pull request merged**